Improve error message when record touch fails.

aguynamedben opened the pull request:
The current error message only indicates that a touch can fail because the record is new. In practice, we saw cases where touches were failing because the record had been destroyed. `persisted?` checks `new_record?` _and_ `destroyed?`. It was confusing to get a message about a new record when in reality we were destroying records.

I also included a helpful tip for users to consider using `persisted?`, `new_record?`, or `destroyed?` before touching.

aguynamedben committed:
Improve error message when record touch fails.

The current error message only indicates that a touch can fail because the record is new. In practice, we saw cases where touches were failing because the record had been destroyed. `persisted?` checks `new_record?` *and* `destroyed?`. It was confusing to get a message about a new record when in reality we were destroying records.

I also included a helpful tip for users to consider using `persisted?`, `new_record?`, or `destroyed?` before touching.

aguynamedben commented:
I'm not sure why the build failed. I don't think it's me.

schneems commented:
I like the extra addition to the error message. Can you try to rebase off of master I think there was a failing test on master that was fixed. Rebasing should fix the problem.

Also style wise the `\` isn't common convention for doing multi line strings. I would suggest using strip_heredoc here is an example https://github.com/rails/rails/blob/b24b9e27510e97ed9878abb2458640f32a48f655/actionpack/lib/action_dispatch/middleware/ssl.rb#L54-L57

aguynamedben commented:
@schneems Cool, I made the change you requested.

* I rebased on master (6a77398)
* I changed the multi-line string to the heredoc style

The relevant tests for activerecord still pass (test/cases/persistence_test.rb and test/cases/touch_later_test.rb). We'll see if the rebase makes the build pass.

aguynamedben commented:
@schneems I think this should be good to go now. Let me know if I should do anything else!

schneems commented:
Thank!
